javatests/arcs/android/entity/TtlHandleTest.kt

```diff
@@ -0,0 +1,202 @@
+        handle2.clear()
+        assertThat(handle2.fetch()).isNull()
+    }
+
+    @Test
+    fun collectionWithExpiredEntities() = runBlockingTest {
+        val time = FakeTime()
+        val handle = createCollectionHandle(time)
+        val entity1 = DummyEntity().apply {
+            num = 1.0
+            texts = setOf("1", "one")
+        }
+        // Set a time in the past. So that this entity is already expired.
+        time.millis = 1L
+        handle.store(entity1)
+        // Then set time to now.
+        time.millis = System.currentTimeMillis()
+
+        val entity2 = DummyEntity().apply {
+            num = 2.0
+            texts = setOf("2", "two")
+        }
+        handle.store(entity2)
+
+        // TODO: after expired entities are filtered out on read, this should only contain entity2.
+        assertThat(handle.fetchAll()).containsExactly(entity1, entity2)
+        
+        // Simulate periodic job triggering.
+        databaseManager.removeExpiredEntities()
+
+        assertThat(handle.fetchAll()).containsExactly(entity2)
+
+        val entity3 = DummyEntity().apply {
+            num = 3.0
+            texts = setOf("3", "three")
+        }
+        handle.store(entity3)
+        assertThat(handle.fetchAll()).containsExactly(entity2, entity3)
+
+        // Creatie a new handle manager with a new storage proxy to confirm entity1 is gone and the
+        // collection is still in a good state.
+        val handle2 = createCollectionHandle(time)
+        
+        assertThat(handle2.fetchAll()).containsExactly(entity2, entity3)
+        
+        val entity4 = DummyEntity().apply {
+            num = 4.0
+            texts = setOf("4", "four")
+        }
+        handle2.store(entity4)
+        assertThat(handle2.fetchAll()).containsExactly(entity2, entity3, entity4)
+    }
+
+    @Test
+    fun sameEntityInTwoCollections() = runBlockingTest {
+        val time = FakeTime()
+        
+        val entity1 = DummyEntity().apply { num = 1.0 }
+        val entity2 = DummyEntity().apply { num = 2.0 }
+        val entity3 = DummyEntity().apply { num = 3.0 }
+        val entity4 = DummyEntity().apply { num = 4.0 }
+
+        val handle1 = createCollectionHandle(time, Ttl.Minutes(1))
+        // A separate collection with the same backing store.
+        val collectionKey2 = ReferenceModeStorageKey(
+            backingKey = backingKey,
+            storageKey = DatabaseStorageKey.Persistent("collection2", DummyEntity.SCHEMA_HASH)
+        )
+        val handle2 = createCollectionHandle(time, Ttl.Minutes(2), collectionKey2)
+        
+        // Entities in handle1 are expired.
+        time.millis = 1L
+        handle1.store(entity1)
+        handle1.store(entity2)
+        
+        // Entities in handle2 are not expired.
+        time.millis = System.currentTimeMillis()
+        handle2.store(entity1)
+        handle2.store(entity3)
+        handle2.store(entity4)
+        
+        time.millis = 1L
```

Comment:
time moving in both directions feels a bit odd...

Reply:
Good point, I have reworked the test to insert all entities at the same time